### PR TITLE
A map-based trie for data indexed by SubCmdIdx

### PR DESCRIPTION
**Pull request by Qining**
Vulkan's new DCE #803 needs a data structure indexed by SubCmdIdx.

### CODE REVIEW

File: gapis/api/subcmd_idx_trie.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		return t.value, true
+	}
+	return nil, false
+}
+
+// Sets a value to the trie with the given SubCmdIdx as index
+func (trie *SubCmdIdxTrie) SetValue(indices SubCmdIdx, v interface{}) {
+	t := trie
+	for _, k := range indices {
+		_, ok := t.nextLevel[k]
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: What I said above is especially true here.

File: gapis/api/subcmd_idx_trie.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+// SubCmdIdxTrie is a map-based trie using SubCmdIdx for indexing the data
+// stored inside.
+type SubCmdIdxTrie struct {
+	hasValue  bool
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: Why do you need `hasValue` and `value`? Are you expecting to have `nil` as a valid value? If not, is there any purpose in the `bool` return value on `GetValue`?

File: gapis/api/subcmd_idx_trie.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	return &SubCmdIdxTrie{
+		hasValue:  false,
+		nextLevel: map[uint64]*SubCmdIdxTrie{},
+	}
+}
+
+// Tries to get the value stored in the trie by the given SubCmdIdx. If a value
+// is stored in the trie for the given SubCmdIdx, returns the value and true,
+// otherwise returns nil and false.
+func (trie *SubCmdIdxTrie) GetValue(indices SubCmdIdx) (interface{}, bool) {
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: Nit: [Drop `Get` from `GetValue`](https://golang.org/doc/effective_go.html#Getters).

File: gapis/api/subcmd_idx_trie.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	t.hasValue = false
+	return true
+}
+
+// Tries to remove a trie node indexed by the given SubCmdIdx. If a node is
+// is found, removes it can all its child nodes, and returns true. Otherwise,
+// returns false.
+func (trie *SubCmdIdxTrie) RemoveNode(indices SubCmdIdx) bool {
+	if len(indices) == 0 {
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: Pointless test - if the `len` were 0, you'd return with the same value.

File: gapis/api/subcmd_idx_trie.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// Tries to remove a value indexed by the given SubCmdIdx in the trie. If a
+// value is found, removes it and returns true. If a value with that SubCmdIdx
+// is not found, returns false. Removing the value from a parent node won't
+// remove the child nodes and their value, i.e.: Removing with [0,1] won't
+// erase the value and the node like: [0, 1, 2].
+func (trie *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
+	t := trie
+	for _, k := range indices {
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: I can't help but feel like you're doing the same traversal logic 3 times here, with slightly different logic once you resolve the node with the given index. How about:

```go
func (t *SubCmdIdxTrie) index(indices SubCmdIdx) *SubCmdIdxTrie {
	var ok bool
	for _, k := range indices {
		if t, ok = t.nextLevel[k]; !ok {
			return nil
		}
	}
	return t
}

func (t *SubCmdIdxTrie) Value(indices SubCmdIdx) interface{} {
	if n := t.index(indices); n != nil {
		return n.value
	}
	return nil
}

func (t *SubCmdIdxTrie) SetValue(indices SubCmdIdx, v interface{}) {
	if n := t.index(indices); n != nil {
		n.value = v
	}
}

func (t *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
	if n := t.index(indices); n != nil {
		t.value = nil
		return true
	}
	return false
}
```

File: gapis/api/subcmd_idx_trie.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		}
+	}
+	t.hasValue = false
+	return true
+}
+
+// Tries to remove a trie node indexed by the given SubCmdIdx. If a node is
+// is found, removes it can all its child nodes, and returns true. Otherwise,
+// returns false.
+func (trie *SubCmdIdxTrie) RemoveNode(indices SubCmdIdx) bool {
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: Continuing with the theme of `nil` == no value, you could simplify the interface for this type by handing `SubCmdIdxTrie.SetValue(i, nil)` as a `RemoveValue(i)`, and automatically remove nodes that have no value or children.

Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: SetValue(i, nil) and RemoveValue(i) are a little bit different. SetValue(i, nil) will create nodes if necessary, but RemoveValue(i) will never create nodes.

Automatically removing nodes is a little bit overkill. This potentially needs to traverse from the bottom to the top of a trie, but right, we can have it that way.

File: gapis/api/subcmd_idx_trie.go (hunk 5 above)
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: The reason I didn't make it this way is that SetValue() will create nodes if necessary, but Value() and Remove() do not modify the Trie and just return failure in such cases. Sure we can have a bool type to control or pass in a callback, but I think either way are not good. These are just three tiny loops, keep it straightforward might be better. But anyway, we can for sure add a 'bool' to the argument list of index() to solve it.

**Comment by Qining** [comment]
Will have traverse() function in the next commit, and traverse() function will have two callbacks for pre-order and post-order traversal. This will enable both
1) Create new nodes during traversing
2) Remove empty nodes during traversing

File: gapis/api/subcmd_idx_trie.go (hunk 5 above)
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: Annoyingly my code snippet is stale - I did update this to account for the creation of nodes, but I have no idea where that comment went. :(

Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]:
```go
func (t *SubCmdIdxTrie) index(indices SubCmdIdx, buildMissing bool) *SubCmdIdxTrie {
	var ok bool
	for _, k := range indices {
		if t, ok = t.nextLevel[k]; !ok {
			if !buildMissing {
				return nil
			}
			next = &SubCmdIdxTrie{}
			t.nextLevel[k], t = next, next
		}
	}
	return t
}

func (t *SubCmdIdxTrie) Value(indices SubCmdIdx) interface{} {
	if n := t.index(indices, false); n != nil {
		return n.value
	}
	return nil
}

func (t *SubCmdIdxTrie) SetValue(indices SubCmdIdx, v interface{}) {
	if n := t.index(indices, true); n != nil {
		n.value = v
	}
}

func (t *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
	if n := t.index(indices, false); n != nil {
		t.value = nil
		return true
	}
	return false
}
```

File: gapis/api/subcmd_idx_trie.go (hunk 6 above)
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: I just scanned #803 to usages of `RemoveNode` and I can't find any. Is it even used?

File: gapis/api/subcmd_idx_trie.go (hunk 2 above)
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: `hasValue` removed in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

File: gapis/api/subcmd_idx_trie.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+// SubCmdIdxTrie is a map-based trie using SubCmdIdx for indexing the data
+// stored inside.
+type SubCmdIdxTrie struct {
+	hasValue  bool
+	value     interface{}
+	nextLevel map[uint64]*SubCmdIdxTrie
+}
+
+// Constructs a new SubCmdIdxTrie instance and returns a pointer to the instance.
```
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: Godocs updated in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

File: gapis/api/subcmd_idx_trie.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+func NewSubCmdIdxTrie() *SubCmdIdxTrie {
+	return &SubCmdIdxTrie{
+		hasValue:  false,
```
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: The initializer is removed in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

File: gapis/api/subcmd_idx_trie.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+		nextLevel: map[uint64]*SubCmdIdxTrie{},
```
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: Fixed in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

File: gapis/api/subcmd_idx_trie.go (hunk 3 above)
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: Changed to `t` in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: Dropped in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

File: gapis/api/subcmd_idx_trie.go (hunk 5 above)
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: hmm, I've updated the `index()` function in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

Now it takes three callback functions. First one for handling the target trie node, second and third are for pre-indexing-order and post-indexing-order callbacks.

So now, we handle can all the requirements for SetValue, RemoveValue

File: gapis/api/subcmd_idx_trie.go (hunk 6 above)
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: No it is not used anywhere, it was added just to make the structure looks more complete. `RemoveNode` has been removed in 12fca8db3ff814a230a9a9fe976420a7e914c4f2.

Now, `RemoveValue()` will take care of deleting the empty nodes.

**Comment by Qining** [comment]
12fca8db3ff814a230a9a9fe976420a7e914c4f2 addes three callbacks function to be called along indexing the `SubCmdIdx` in the `trie`.

All comments should have been resolved.

File: gapis/api/subcmd_idx_trie.go (hunk 4 above)
Comment (Qining) [inline gapis/api/subcmd_idx_trie.go]: removed in 12fca8db3ff814a230a9a9fe976420a7e914c4f2

File: gapis/api/subcmd_idx_trie.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+			}
+		}, nil)
+}
+
+// RemoveValue tries to remove a value indexed by the given SubCmdIdx in the
+// trie. If a value is found, removes it and returns true. If a value with that
+// SubCmdIdx is not found, returns false.
+func (t *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
+	hadValue := false
+	r := t.index(indices, func(t *SubCmdIdxTrie) {
```
Comment (ben-clayton) [inline gapis/api/subcmd_idx_trie.go]: Due to the closures, captured value, order of calls, I find this exceptionally hard to read and follow.
Here's an alternative implementation that passes the tests:

```go

// SubCmdIdxTrie is a map-based trie using SubCmdIdx for indexing the data
// stored inside.
type SubCmdIdxTrie struct {
	value    interface{}
	children map[uint64]*SubCmdIdxTrie
}

// Value returnes the value stored in the trie indexed by the given SubCmdIdx.
// if no value is found by the given SubCmdIdx, returns nil.
func (t *SubCmdIdxTrie) Value(indices SubCmdIdx) interface{} {
	for t != nil && len(indices) > 0 && len(t.children) > 0 {
		t, indices = t.children[indices[0]], indices[1:]
	}
	if t == nil || len(indices) > 0 {
		return nil // invalid index
	}
	return t.value
}

// SetValue sets a value to the trie with the given SubCmdIdx as index.
func (t *SubCmdIdxTrie) SetValue(indices SubCmdIdx, v interface{}) {
	if v == nil { // nil value behaves the same as no value
		t.RemoveValue(indices)
		return
	}
	for _, i := range indices {
		next, ok := t.children[i]
		if !ok { // child does not exist - create it
			if t.children == nil {
				t.children = map[uint64]*SubCmdIdxTrie{}
			}
			next = &SubCmdIdxTrie{}
			t.children[i] = next
		}
		t = next
	}
	t.value = v
}

// RemoveValue tries to remove a value indexed by the given SubCmdIdx in the
// trie. If a value is found, removes it and returns true. If a value with that
// SubCmdIdx is not found, returns false.
func (t *SubCmdIdxTrie) RemoveValue(indices SubCmdIdx) bool {
	n := len(indices)
	stack := make([]*SubCmdIdxTrie, n+1)
	stack[0] = t
	for i, j := range indices { // build a stack of nodes starting with t
		t = t.children[j]
		if t == nil {
			return false // invalid index
		}
		stack[i+1] = t
	}
	if t.value == nil {
		return false // cannot remove a value it doesn't have
	}
	t.value = nil             // clear the value
	for i := n; i >= 0; i-- { // remove nodes that have no value or children
		if t.value != nil || len(t.children) > 0 {
			break // node cannot be deleted
		}
		delete(stack[i].children, indices[i-1])
		t = stack[i]
	}
	return true
}
```

This implementation is fractionally faster too:

Old:
```bash
$ go test ./gapis/api/ --bench=.                                                                                                                          
BenchmarkSubCmdIdxTrie-48         100000             11770 ns/op 
```

New:
```bash
$ go test ./gapis/api/ --bench=.                                                                                                                                                    
BenchmarkSubCmdIdxTrie-48         200000             10194 ns/op
```

**Comment by Qining** [comment]
@ben-clayton Updated.

**Pull request merged**